go/ephemeral/lib.go

```diff
@@ -0,0 +1,126 @@
+package ephemeral
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+type EKLib struct {
+	libkb.Contextified
+	sync.Mutex
+}
+
+func NewEKLib(g *libkb.GlobalContext) *EKLib {
+	return &EKLib{
+		Contextified: libkb.NewContextified(g),
+	}
+}
+
+func (e *EKLib) KeygenIfNeeded(ctx context.Context) (err error) {
+	defer e.G().CTrace(ctx, "KeygenIfNeeded", func() error { return err })()
+	e.Lock()
+	defer e.Unlock()
+
+	if loggedIn, err := e.G().LoginState().LoggedInLoad(); err != nil {
+		return err
+	} else if !loggedIn {
+		return fmt.Errorf("Aborting ephemeral key generation, user is not logged in!")
+	}
+
+	merkleRootPtr, err := e.G().GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
+	if err != nil {
+		return err
+	}
+	merkleRoot := *merkleRootPtr
+
+	deviceEKNeeded, err := e.newDeviceEKNeeded(ctx, merkleRoot)
+	if err != nil {
+		return err
+	}
+	if deviceEKNeeded {
+		_, err = publishNewDeviceEK(ctx, e.G(), merkleRoot)
+		if err != nil {
+			return err
+		}
+	}
+
+	userEKNeeded, err := e.newUserEKNeeded(ctx, merkleRoot)
+	if err != nil {
+		return err
+	}
+
+	if userEKNeeded {
+		_, err = publishNewUserEK(ctx, e.G(), merkleRoot)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Can we add some TODO's here about making the device and user uploads atomic? We did that whiteboard example about how you can get into trouble if you publish a new deviceEK but fail to publish a new userEK. The short version was that the new deviceEK cuts off the extension window for the old key (by design), but without a new userEK you still have new teamEKs being created that only the old deviceEK can ultimately decrypt.

Reply:
left a note to myself to discuss further in person